cognite/client/utils/_concurrency.py

```diff
@@ -109,38 +110,108 @@ def collect_exc_info_and_raise(
+
+
+class _Executor(Protocol):
+    def submit(self, fn: Callable[..., T_Result], *args: Any, **kwargs: Any) -> _Future[T_Result]:
+        ...
+
+
+class _Future(Protocol[T_Result]):
+    def result(self) -> T_Result:
+        ...
+
+
+class SyncFuture(_Future):
+    def __init__(self, fn: Callable[..., T_Result], *args: Any, **kwargs: Any):
+        self.__fn = fn
+        self.__args = args
+        self.__kwargs = kwargs
+
+    def result(self) -> T_Result:
+        return self.__fn(*self.__args, **self.__kwargs)
+
+
+class MainThreadExecutor(_Executor):
+    """
+    In order to support executing sdk methods in the browser using pyodide (a port of CPython to webassembly),
+    we need to be able to turn off the usage of threading. So we have this executor which implements the Executor
+    protocol but just executes everything serially in the main thread.
+    """
+
+    def submit(self, fn: Callable, *args: Any, **kwargs: Any) -> SyncFuture:
+        return SyncFuture(fn, *args, **kwargs)
+
+
+_THREAD_POOL_EXECUTOR_SINGLETON: ThreadPoolExecutor
+_MAIN_THREAD_EXECUTOR_SINGLETON = MainThreadExecutor()
+
+
+def _shutdown_thread_pool_executor() -> None:
+    try:
+        tpe = _THREAD_POOL_EXECUTOR_SINGLETON
+        tpe.shutdown(wait=False)
+    except NameError:
+        # TPE has not been initialized
+        pass
+
+
+atexit.register(_shutdown_thread_pool_executor)
+
+
+class ConcurrencySettings:
+    executor_type: Literal["threadpool", "mainthread"] = "threadpool"
+
+
+def get_executor(max_workers: int) -> _Executor:
+    global _THREAD_POOL_EXECUTOR_SINGLETON
+
     if max_workers < 1:
         raise RuntimeError(f"Number of workers should be >= 1, was {max_workers}")
 
-    with ThreadPoolExecutor(max_workers) as p:
-        futures = []
-        for task in tasks:
-            if isinstance(task, dict):
-                futures.append(p.submit(func, **task))
-            elif isinstance(task, tuple):
-                futures.append(p.submit(func, *task))
-
-        successful_tasks = []
-        failed_tasks = []
-        unknown_result_tasks = []
-        results = []
-        exceptions = []
-        for i, f in enumerate(futures):
-            try:
-                res = f.result()
-                successful_tasks.append(tasks[i])
-                results.append(res)
-            except Exception as e:
-                exceptions.append(e)
-                if isinstance(e, CogniteAPIError):
-                    if e.code < 500:
-                        failed_tasks.append(tasks[i])
-                    else:
-                        unknown_result_tasks.append(tasks[i])
-                else:
+    if ConcurrencySettings.executor_type == "threadpool":
+        try:
+            executor: _Executor = _THREAD_POOL_EXECUTOR_SINGLETON
+        except NameError:
+            # TPE has not been initialized
+            executor = _THREAD_POOL_EXECUTOR_SINGLETON = ThreadPoolExecutor(max_workers)
```

Comment:
Is it really a singleton? ;P
```suggestion
            executor = _THREAD_POOL_EXECUTOR_SINGLETON = _THREAD_POOL_EXECUTOR_SINGLETON(max_workers)
```

Reply:
hmm, yes?

Reply:
note that l146 does not assign the type ThreadPoolExecutor to _THREAD_POOL_EXECUTOR_SINGLETON, it just declares the variable with a type annotation.

Reply:
# 🤯 
Oh damn, you're right... And I read those lines several times without noticing